File: libs/headless-figma-clone/src/eval/gates.ts
import type { FileEnvelope } from '../model/types.js';
import type { EditGraph, EvalGates, SubCheckResult } from './types.js';
import { descendantIds, nodeExists } from './tree.js';

function gateResult(id: string, score: number, details?: Record<string, unknown>): SubCheckResult {
  return { id, category: 'gates', score, applicable: true, weight: 1, details };
}

export function runGates(
  _before: FileEnvelope,
  after: FileEnvelope,
  graph: EditGraph,
  gates: EvalGates | undefined
): SubCheckResult[] {
  if (!gates) return [];

  const results: SubCheckResult[] = [];

  if (gates.require_change) {
    results.push(
      gateResult('gates.require_change', graph.equal ? 0 : 1, { equal: graph.equal })
    );
  }

  for (const id of gates.preserve_ids ?? []) {
    const ok = nodeExists(after, id);
    results.push(gateResult(`gates.preserve.${id}`, ok ? 1 : 0, { node_id: id }));
  }

  for (const id of gates.forbid_delete_ids ?? []) {
    const deleted = graph.deletedIds.has(id);
    results.push(gateResult(`gates.forbid_delete.${id}`, deleted ? 0 : 1, { node_id: id }));
  }

  if (gates.max_change_outside_ids?.length) {
    const allowed = new Set<string>();
    for (const rootId of gates.max_change_outside_ids) {
      for (const id of descendantIds(after, rootId)) allowed.add(id);
    }

    const changedOutside: string[] = [];
    for (const id of graph.addedIds) {
      if (!allowed.has(id)) changedOutside.push(id);
    }
    for (const id of graph.modifiedIds) {
      if (!allowed.has(id)) changedOutside.push(id);
    }
    for (const id of graph.deletedIds) {
      if (!allowed.has(id)) changedOutside.push(id);
    }

    const totalChanged = graph.addedIds.size + graph.modifiedIds.size + graph.deletedIds.size;
    const outside = changedOutside.length;
    const locality = totalChanged === 0 ? 1 : Math.max(0, 1 - outside / totalChanged);
    results.push(
      gateResult('gates.max_change_outside', locality, {
        changed_outside: outside,
        total_changed: totalChanged,
      })
    );
  }

  return results;
}

File: envs/figma-design/tasks/with-metadata/tests/figma_eval/gates.py
from __future__ import annotations

from typing import Any

from .edit_graph import EditGraph
from .tree import descendant_ids, is_node_under_roots, node_exists
from .types import Envelope, SubCheckResult


def _gate_result(gate_id: str, score: float, details: dict[str, Any] | None = None) -> SubCheckResult:
    return SubCheckResult(
        id=gate_id,
        category="gates",
        score=score,
        applicable=True,
        weight=1.0,
        details=details,
    )


def _allowed_before_scope(before: Envelope, root_ids: list[str]) -> set[str]:
    allowed: set[str] = set()
    for root_id in root_ids:
        allowed |= descendant_ids(before, root_id)
    return allowed


def _has_change_outside_allowed_region(
    graph: EditGraph,
    before: Envelope,
    after: Envelope,
    root_ids: list[str],
) -> bool:
    allowed_before = _allowed_before_scope(before, root_ids)

    for nid in graph.added_ids:
        if not is_node_under_roots(after, nid, root_ids):
            return True

    for nid in graph.modified_ids:
        if nid not in allowed_before or not is_node_under_roots(after, nid, root_ids):
            return True

    for nid in graph.deleted_ids:
        if nid not in allowed_before:
            return True

    return False


def run_gates(
    before: Envelope,
    after: Envelope,
    graph: EditGraph,
    gates: dict[str, Any] | None,
) -> list[SubCheckResult]:
    if not gates:
        return []

    results: list[SubCheckResult] = []

    if gates.get("require_change"):
        results.append(
            _gate_result("gates.require_change", 0.0 if graph.equal else 1.0, {"equal": graph.equal})
        )

    for nid in gates.get("preserve_ids") or []:
        ok = node_exists(after, nid)
        results.append(_gate_result(f"gates.preserve.{nid}", 1.0 if ok else 0.0, {"node_id": nid}))

    allowed_inside = gates.get("allowed_change_inside_ids")
    if allowed_inside:
        outside = _has_change_outside_allowed_region(graph, before, after, allowed_inside)
        results.append(
            _gate_result(
                "gates.allowed_change_inside",
                0.0 if outside else 1.0,
                {"allowed_roots": allowed_inside},
            )
        )

    if gates.get("additions_only"):
        ok = not graph.modified_ids and not graph.deleted_ids
        results.append(_gate_result("gates.additions_only", 1.0 if ok else 0.0))

    return results
